### PR TITLE
[SPARK-36531][SPARK-36515][PYTHON] Improve test coverage for data_type_ops/* and groupby.

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes improving test coverage for pandas-on-Spark data types & GroupBy code base, which is written in `data_type_ops/*.py` and `groupby.py` separately.

This PR did the following to improve coverage:
- Add unittest for untested code
- Fix unittest which is not tested properly
- Remove unused code

**NOTE**: This PR is not only include the test-only update, for example it includes the fixing `astype` for binary ops.

pandas-on-Spark Series we have:
```python
>>> psser
0    [49]
1    [50]
2    [51]
dtype: object
```

before:
```python
>>> psser.astype(bool)
Traceback (most recent call last):
...
pyspark.sql.utils.AnalysisException: cannot resolve 'CAST(`0` AS BOOLEAN)' due to data type mismatch: cannot cast binary to boolean;
...
```

after:
```python
>>> psser.astype(bool)
0    True
1    True
2    True
dtype: bool
```

### Why are the changes needed?

To make the project healthier by improving coverage.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unittest.

### CODE REVIEW

File: python/pyspark/pandas/data_type_ops/binary_ops.py [hunk 1]
```diff
@@ -100,7 +99,9 @@ def astype(self, index_ops: IndexOpsLike, dtype: Union[str, type, Dtype]) -> Ind
         if isinstance(dtype, CategoricalDtype):
             return _as_categorical_type(index_ops, dtype, spark_type)
         elif isinstance(spark_type, BooleanType):
-            return _as_bool_type(index_ops, dtype)
+            # Cannot cast binary to boolean in Spark.
+            # We should cast binary to str first, and cast it to boolean
+            return index_ops.astype(str).astype(bool)
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/data_type_ops/binary_ops.py:104]: @itholic can you describe the changes that are not test-only in pr description?

Comment (HyukjinKwon) [reply python/pyspark/pandas/data_type_ops/binary_ops.py:104]: cc @xinrong-databricks FYI

**Comment by HyukjinKwon** [comment]
looks fine. cc @xinrong-databricks and @ueshin

**Review by HyukjinKwon — COMMENTED** [review]
looks fine. cc @xinrong-databricks and @ueshin

File: python/pyspark/pandas/data_type_ops/binary_ops.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/pandas/data_type_ops/binary_ops.py:104]: added!

File: python/pyspark/pandas/tests/data_type_ops/test_datetime_ops.py [hunk 2]
```diff
@@ -191,10 +191,11 @@ def test_astype(self):
         pser = self.pser
         psser = self.psser
         self.assert_eq(pser.astype(str), psser.astype(str))
-        self.assert_eq(pser.astype("category"), psser.astype("category"))
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_datetime_ops.py:194]: Why is this removed?

File: python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py [hunk 3]
```diff
@@ -190,6 +190,7 @@ def test_astype(self):
         self.assert_eq(pser.astype(str), psser.astype(str))
         self.assert_eq(pser.astype(bool), psser.astype(bool))
         self.assert_eq(pser.astype("category"), psser.astype("category"))
+        self.assert_eq(pser.astype("category"), psser.astype("category"))
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:193]: This is duplicated with the previous one?

File: python/pyspark/pandas/tests/data_type_ops/test_string_ops.py [hunk 4]
```diff
@@ -24,7 +24,9 @@
 from pyspark import pandas as ps
 from pyspark.pandas.config import option_context
 from pyspark.pandas.tests.data_type_ops.testing_utils import TestCasesUtils
-from pyspark.pandas.typedef.typehints import extension_object_dtypes_available
+from pyspark.pandas.typedef.typehints import (
+    extension_object_dtypes_available,
+)
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_string_ops.py:29]: This is an unrelated change?

Comment (itholic) [reply python/pyspark/pandas/tests/data_type_ops/test_string_ops.py:29]: reverted!

File: python/pyspark/pandas/data_type_ops/binary_ops.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/pandas/data_type_ops/binary_ops.py:104]: can you describe in more details with the input/output before/after this change.

Comment (HyukjinKwon) [reply python/pyspark/pandas/data_type_ops/binary_ops.py:104]: btw, why not fixing it in `_as_bool_type`?

Comment (itholic) [reply python/pyspark/pandas/data_type_ops/binary_ops.py]: moved it into `_as_bool_type` and updated the PR description!

Comment (ueshin) [reply python/pyspark/pandas/data_type_ops/binary_ops.py]: `_as_bool_type` should not include `index_ops`'s type dependent code.
We should revert the change c429108d380d5dfaf4402d8382bb26d6a0ca32fc.

**Review by ueshin — APPROVED** [review]
LGTM.
